### PR TITLE
API Add support for BasicAuth to be used for native permission checks

### PR DESCRIPTION
**Pull request by robbieaverill**
The BasicAuth implementation will adequately protect the API endpoint, and will validate against a Member, but (by design) it will not be stored in the session and as such cannot be used for native SilverStripe permission checking (`Page::canView`).

This change adds an `AuthenticatorInterface` which can be implemented by various authenticator classes to use the `HTTPRequest` provided to GraphQL to authenticate against members, and return a `Member` instance if successful.

This PR includes a `BasicAuthAuthenticator`.

Example configuration:

```yaml
SilverStripe\GraphQL:
  # Enforce HTTP basic authentication for GraphQL requests
  authenticators:
    - class: SilverStripe\GraphQL\Auth\BasicAuthAuthenticator
      priority: 10
```

The only non-obvious implication here is that even if you have a global basic auth requirement set on your site, you'll still need to configure the `BasicAuthAuthenticator` for GraphQL.

The docs have been updated with authentication information, a configuration example, and an example for GraphiQL (app only) client implementation using the base64 encoded header.

---

Starts #6

### CODE REVIEW

**Comment by robbieaverill** [comment]
Dear Codecov.io - I can't easily test those uncovered lines, because the BasicAuth class is called from the parent of the GraphQL controller and can't be mocked.

I don't like this change a whole lot anyway, so I'm not super willing to write a long winded test for functionality that is tested elsewhere anyway. :)

Best, Robbie

File: README.md [hunk 1]
```diff
@@ -457,6 +457,62 @@ This tool is available in **dev mode only**. It can be accessed at
...
+are performing requests from an anonymous/external application you may need to
+authenticate before you can complete a request.
+
+Please note that when implementing GraphQL resources it is the developer's
+responsibility to ensure that permission checks are implemented wherever
+resources are accessed.
+
+### Basic Authentication
+
+Silverstripe has built in support for basic authentication, and it can be configured
```
Comment (chillu) [inline README.md]: I'd call it "HTTP basic authentication", and link to the wikipedia entry: https://en.wikipedia.org/wiki/Basic_access_authentication. Too easy to misread as "some non-descript form of non-advanced authentication" otherwise :D

File: README.md [hunk 2]
```diff
@@ -457,6 +457,62 @@ This tool is available in **dev mode only**. It can be accessed at
...
+does not support custom HTTP headers at this point.
+
+Your custom header should follow the following format:
+
+```
+# Key: Value
+Authorization: Basic aGVsbG86d29ybGQ=
+```
+
+`Basic` is followed by a [base64 encoded](https://en.wikipedia.org/wiki/Base64)
```
Comment (chillu) [inline README.md]: I'd just add the `bash` example from the wikipedia page here, for ease of use.

Comment (chillu) [inline README.md]: And add a note that authentication is transferred in plain text via this method, so we strongly recommend using TLS for non-development use.

**Comment by robbieaverill** [comment]
@chillu as per our chat on Slack I've re-implemented the authentication structure to be less focussed on the controller.

Now, the controller calls the "auth handler" to decide if any authentication requirements need to be met.

The Handler will load a configurable list of authenticators in a priority order and call the first one's `authenticate` method. If successfully authenticated it will set the Member returned to the Manager for use in the context later, or throw an exception on failure.

Authenticators themselves will implement the `AuthenticatorInterface` and will be given the current request object if they need it.

This is a bit more future proof, and will allow us to plug in an OAuth 2 interface more easily.

**Comment by robbieaverill** [comment]
**Note:** squash on merge and set a nice commit message :-) or I'll rebate first. The previous commits are left here for brevity.

**Comment by robbieaverill** [comment]
## Use case - for extensibility (example only)

mysite configuration:
```yml
SilverStripe\GraphQL:
  authenticators:
    - class: MyProject\GraphQL\DevOnlyAuthenticator
      priority: 100
```

Authenticator:
```php
<?php

namespace MyProject\GraphQL;

use SilverStripe\Control\HTTPRequest;
use SilverStripe\GraphQL\Auth\AuthenticatorInterface;
use SilverStripe\Security\Member;

class DevOnlyAuthenticator implements AuthenticatorInterface
{
    public function authenticate(HTTPRequest $request)
    {
        return Member::get()->filter('Email', $request->getHeader('Email'))->first();
    }
}
```

Then add `Email: my-cms@email-address.com` as a custom header in the GraphiQL app.

File: src/Auth/Handler.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+     * @var Manager
+     */
+    protected $manager;
+
+    /**
+     * @param Manager
+     */
+    public function __construct($manager)
+    {
+        $this->setManager($manager);
```
Comment (chillu) [inline src/Auth/Handler.php]: Making `Manager` a dependency in this class seems a bit overkill - what's the reasoning? I would've just returned the member from `requireAuthentication()` and set the member in `Controller`? You can infer authentication success from the presence of a member, same as with `BasicAuth::requireLogin()`

Comment (robbieaverill) [inline src/Auth/Handler.php]: Yep- good call

Comment (robbieaverill) [inline src/Auth/Handler.php]: @chillu much simpler without it! Thanks for pointing that out. Update pushed.

**Comment by robbieaverill** [comment]
* Squashed

File: src/Controller.php [hunk 4]
```diff
@@ -43,9 +48,13 @@ public function index(HTTPRequest $request)
             $variables = json_decode($variables, true);
         }
 
-        $manager = $this->getManager();
+        $this->setManager($manager = $this->getManager());
```
Comment (chillu) [inline src/Controller.php]: Why was this necessary? `setManager()` doesn't do anything but `$this->manager = $manager`, which is already done in `getManager()`?
